Implement rplidar as camera subtype

kkufieta opened the pull request:
@jeremyhyde-viam Implemented rplidar as a camera sybtype.
@kkufieta Made it pretty & fixed a few bugs.

Otterverse commented on cmd/server/main.go:
```diff
@@ -39,80 +38,64 @@ type Arguments struct {
-		golog.Global.Debugf("detected %d lidar devices", len(lidarComponents))
-		for _, comp := range lidarComponents {
+	// Check if USB device is available
+	// TODO(jeremy): add search filter for product and model info to confirm rplidar is present instead of assuming
```
This TODO is either already done it looks like (so comment needs to be removed) or misunderstanding something. usb.SearchFilter{} is ignored on linux, so wouldn't work anyway, the vendorID/productID function is the only one that matters there.

jeremyhyde-viam replied:
When redoing this repo I needed to alter the searching function being used as search.Devices was no longer available. During development I turned the vendorID/productID filter off (see below that the function just returns true). @Otterverse should we re-add the filter for the PR?

Otterverse replied:
Vendor/Product should be enough to always get the right device and it'd be much better to use the device file returned from that rather than hardcoding a generic device name like /dev/ttyUSB0. The actual "SearchFilter" is mac only, and if anything may need to be expanded (not narrowed) to include additional device classes if this doesn't show up as a typical serial/dialup device under Mac.

kkufieta replied:
I don't actually understand this part - can you walk me through it so I can change this? E.g. a code-snippet or solution would work :)

kkufieta replied:
never mind - I implemented this using the search functionality and it seems to work well so far.

kkufieta replied:
Done.

Otterverse commented on cmd/server/main.go:
```diff
@@ -4,30 +4,29 @@ import (
-	robotimpl "go.viam.com/core/robot/impl"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/metadata/service"
+	"go.viam.com/rdk/rlog"
```
You shouldn't need this AND golog (above) at the same time. This literally just pulls in golog. (This applies to the other files that do this as well.)

kkufieta replied:
In this case, `ContextualMain` requires a `golog.Logger`. I tried to recreate that using only the `rlog` module, e.g. by doing `rlog.golog.Logger`, but it wouldn't compile. I didn't find any examples that reflect solve it using only `rlog` or `golog`, either.

Do you know how to fix this, or if this is even possible without importing both `golog` and `rlog`?

kkufieta replied:
Fixed as discussed in chat; am using mix of both, but am making sure that all `golog.Global` are changed to `rlog.Logger`.

Done.

Otterverse commented on cmd/server/main.go:
```diff
@@ -39,80 +38,64 @@ type Arguments struct {
 	}
+
 	if argsParsed.Port == 0 {
+		golog.Global.Debugf("using default port %d ", defaultPort)
```
There is a logger being passed to this function. It should be used here instead of directly accessing a global.

kkufieta replied:
Done.

Otterverse commented on cmd/server/main.go:
```diff
@@ -39,80 +38,64 @@ type Arguments struct {
+	if len(usbDevices) != 0 {
+		golog.Global.Debugf("detected %d lidar devices", len(usbDevices))
+		for _, comp := range usbDevices {
 			golog.Global.Debug(comp)
```
As before, these loggers instances should use the logger that was passed to the function.

kkufieta replied:
Done.

Otterverse commented on cmd/server/main.go:
```diff
@@ -39,80 +38,64 @@ type Arguments struct {
+		Name:       "rplidar",
+		Type:       config.ComponentTypeCamera,
+		Model:      rplidar.ModelName,
+		Attributes: config.AttributeMap{"device_path": "/dev/ttyUSB0"},
```
Why is the device path hard coded here, when above there's a USB search above for the right vendor and product id? Does the result from that not work for some reason (I know there were some issues on Mac, but...)

kkufieta replied:
Fixed it to use the search result! Thanks for pointing it out.

Done.

Otterverse commented on configs/rplidar_process.json:
```diff
@@ -0,0 +1,16 @@
+        {
+            "id": "rplidar_process",
+            "name": "/home/jeremyhyde-viam/rplidar/main",
+            "cwd": "/home/jeremyhyde-viam/rplidar",
```
Personal home directories shouldn't be in config files or code, as it won't work for anyone else (and it's bad practice to "leak" personal data like names.)

jeremyhyde-viam replied:
probably bet to remove config directory and files for this PR

kkufieta replied:
Removed. Done.

Otterverse commented on gen/third_party/rplidar_sdk-release-v1.12.0/sdk/sdk/src/rplidar_driver.cpp:
```diff
@@ -1821,7 +1821,7 @@ u_result RPlidarDriverImplCommon::grabScanData(rplidar_response_measurement_node
 u_result RPlidarDriverImplCommon::grabScanDataHq(rplidar_response_measurement_node_hq_t * nodebuffer, size_t & count, _u32 timeout)
 {
-    switch (_dataEvt.wait(timeout))
+    switch (int(_dataEvt.wait(timeout)))
```
I believe these should no longer need a cast here since the actual function signature was changed in event.h.

kkufieta replied:
Done.

Otterverse commented on rplidar.go:
```diff
@@ -1,36 +1,55 @@
+)
+
+const (
+	// ModelName is how the lidar will be registered into core.
```
[nit] "core" is "RDK" now.

Otterverse commented on rplidar.go:
```diff
@@ -149,21 +168,34 @@ func NewDevice(devicePath string) (*Device, error) {
+	err := d.Start()
+	if err != nil {
+		driver = nil
+		gen.RPlidarDriverDisposeDriver(driver)
```
I think this is backwards. You need to dispose first, before setting it to nil (as is done in the other similar cases above.)

kkufieta replied:
ugh sorry this is a mess! Shouldn't even be here!

kkufieta replied:
Taking that back - it belongs here - but you're right, the order has to be reversed.

Done.

Otterverse commented on rplidar.go:
```diff
@@ -249,86 +286,83 @@ func (d *Device) filterParams() (minAngleDiff float64, maxDistDiff float64) {
-		//nolint
-		d.scan(lidar.ScanOptions{Count: 10})
+		d.scan(ctx)
 		time.Sleep(time.Second)
```
This is a long sleep without checking context. Should use selectContextOrWait() here instead.

kkufieta replied:
Done.

Otterverse commented on rplidar.go:
```diff
@@ -348,40 +382,50 @@ func (d *Device) scan(options lidar.ScanOptions) (lidar.Measurements, error) {
+
+	t_now := (time.Now())
+	t_now2 := t_now.Format(time.RFC3339)
+	t_str := strings.Replace(strings.Replace(t_now2, ":", "_", 3), "-", "", 2)
```
Several things here. First, don't use snake_case in go, use camelCase. Second, no need for the parentheses around time.Now(). Third, go's time formatting is entirely customizable. Fourth, methods are composable here. Thus the exact same output can be had in one line:

`tStr := time.Now().Format("20060102T15_04_05Z")`

(Go uses a specific fixed date/time as the way to format things... this isn't actually hardcoding 2006)

jeremyhyde-viam replied:
Interesting! Good to know James, I've always found string to time conversions a hassle and this seems a lot better

kkufieta replied:
Done.

kkufieta replied:
Actually - @Otterverse I am also trying to capture fractional seconds. I tried this:

`timeStamp.Format("20060102T15_04_05_00")`, but it doesn't seem to work for the fractional seconds. Can you see anything obviously wrong with this? And, in your example, what does the `Z` stand for?

Otterverse replied:
Z indicates "Zulu Time" (aka, UTC) in many timestamp formats, including RFC3339. I think I may have misunderstood golang's formatting slightly. I was adding the Z by itself, when really I needed to specify "Z0700" or "ZO7" or "Z07:00". It'll print the hours and minutes of the timezone offset for most timezones, but if the system is set to UTC, then it'll just print "Z"

Adding fractional seconds is just adding a decimal after the minutes and appending some zeros.

E.g. with four decimals: `Format("2006-01-02T15_04_05.0000Z0700")`
Right now on my PC (at UTC -0800) that'll print: `2022-02-01T18_31_12.2556-0800`

Go's time formatting is here: https://pkg.go.dev/time#pkg-constants and it's a little confusing as I mentioned. The problem is that you're trying to replace dashes with "" so that will break things potenailly, as there'd be no difference between UTC+8 and UTC-8 timezones. If you just want a usable filename that sorts in order, then force UTC and don't print the zone.

`Format("2006-01-02T15_04_05.0000")` >>> `2022-02-02T02_36_27.7745`

Otterverse commented on rplidar.go:
```diff
@@ -348,40 +382,50 @@ func (d *Device) scan(options lidar.ScanOptions) (lidar.Measurements, error) {
-			filteredMeasurements = append(filteredMeasurements, curr)
-		}
-		detectedRay = false
+	defer utils.UncheckedErrorFunc(f.Close)
```
So the notes on this UncheckedErrorFunc (in goutils) explcitly say "never use this for closing writers" and "f" is is definitely a file writer.

https://github.com/viamrobotics/goutils/blob/ee8bd5e1fc8bdc0e3449e936988a6d2f3d7da508/error.go#L72

This close should be handled explicitly and checked for errors.

kkufieta replied:
I stumbled also over that comment - but thought that `f` is a file, whereas `w` is a writer. Am I mistaken?

Otterverse replied:
In Go parlance, anything that supports the "io.Writer" interface is a "writer." That definitely includes files, but a lot of other stuff too. Files are a particularly bad case though, because Write() isn't always guaranteed to be on disk until after the file is closed. So if you did a bunch of writes (with no error) the Close() itself can return an error about "can't write to disk" or similar, and even the previous Write() opertions may be lost.

If you're curious, a much more in depth explanation is here: https://www.joeshaw.org/dont-defer-close-on-writable-files/

kkufieta replied:
Thanks for explaining and for the link!

Done.

Otterverse reviewed (changes requested):
Mostly a lot of small details that need fixing. If you have questions, feel free to ping me.

In addition to the individual notes, I want to note that bar.las (a binary file, so I can't comment directly on it) likely shouldn't be included in the git commit (unless I'm mistaken and is is required/static.) If this is something commonly generated here as output, may be a good idea to put it in .gitignore.

jeremyhyde-viam commented on cmd/client/main.go:
```diff
@@ -35,36 +38,26 @@ func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) error
 
-	ticker := time.NewTicker(time.Second)
+	// Run loop
+	ticker := time.NewTicker(5 * time.Second)
```
This is malleable, I chose 5s originally  to allow .las files not to be generated too quickly. My opinion is we stick with 5s for now until we need to speed it up (potentially for slam or for other applications)

kkufieta replied:
ok, will keep for now. 

Done.

kkufieta replied:
Actually, taking this back up. Should I use `selectContextOrWait` instead of the ticker here, @Otterverse ?

Otterverse replied:
Actually, yeah, probably so. Good catch!

jeremyhyde-viam commented on cmd/server/main.go:
```diff
@@ -39,80 +38,64 @@ type Arguments struct {
-	return rpcServer.Serve(listener)
+	options := web.NewOptions()
+	options.Network = config.NetworkConfig{BindAddress: fmt.Sprintf(":%d", port)}
+	return webserver.RunWeb(ctx, myRobot, options, logger)
```
Note: Kat experienced issues closing out of the server needing to resort to CTRL-\ to do it. Since the server is basically be held by the webserver.RunWeb at this point, my thought that this indicates a potential issue with it and not this cmd/server/main.go repo. But I could be mistaken.

kkufieta replied:
The issue was that there was a workgroup added to the code that:
a) was not needed
b) was blocking the `Close` function from finishing.

jeremyhyde-viam commented on configs/rplidar_remote.json:
```diff
@@ -0,0 +1,8 @@
+    "remotes": [
+        {
+            "name": "rplidar",
+            "address": "localhost:8081"
```
See above

kkufieta replied:
Done.

jeremyhyde-viam commented on rplidar.go:
```diff
@@ -1,36 +1,55 @@
-		return NewDevice(config.Host)
-	})
-	lidar.RegisterType(rplidar.Type, lidar.TypeRegistration{
-		USBInfo: &usb.Identifier{
```
USBInfo (used in search device?) was performed here but removed in update

kkufieta replied:
Does that mean it has to be removed in general?

kkufieta replied:
nvm - put it outside of the init function and am using it for the search for the device.

jeremyhyde-viam commented on rplidar.go:
```diff
@@ -249,86 +286,83 @@ func (d *Device) filterParams() (minAngleDiff float64, maxDistDiff float64) {
 }
 
 // Bounds returns the square meter bounds of the device.
-func (d *Device) Bounds(ctx context.Context) (r2.Point, error) {
```
bounds is potentially not needed anymore (but still might be useful). It was a class function of lidar but no longer exists in camera

kkufieta replied:
I'll keep it around for now. Might remove it in a future iteration of cleaning up rplidar, with help of Eric.

kkufieta commented on rplidar.go:
```diff
@@ -249,86 +286,83 @@ func (d *Device) filterParams() (minAngleDiff float64, maxDistDiff float64) {
 		d.scannedOnce = true
 		// discard scans for warmup
-		//nolint
-		d.scan(lidar.ScanOptions{Count: 10})
```
@jeremyhyde-viam Is there a reason why this was removed? Do I assume correctly, that the original intention was to skip the first 10 scans, since they're garbage?

kkufieta replied:
Got it. I changed the way this was implemented, but kept the warmup routine in the scan function for two reasons:
1. I assume Eric had a good reason to put it in there, most likely after realizing it's necessary. If we want to get rid of it, we should probably chat with him first.
2. Before being able to scan, there's a whole setup involved. It's easiest for now to keep this where it is.

Done.

jeremyhyde-viam commented on rplidar.go:
```diff
@@ -395,3 +439,64 @@ func (d *Device) AngularResolution(ctx context.Context) (float64, error) {
+func (d *Device) Close(ctx context.Context) error {
+	if d.driver != nil {
+		d.Stop()
+		gen.RPlidarDriverDisposeDriver(d.driver)
```
Great catch Kat!

jeremyhyde-viam reviewed (changes requested):
Primarily looks good, James commented on most issues, most of my comments are optional (such as removing Bounds) but I'd like to see the final state before final approval

kkufieta commented on cmd/client/main.go:
```diff
@@ -78,13 +71,13 @@ func runClient(ctx context.Context, deviceAddress string, logger golog.Logger) (
 		}
 
-		measurements, err := lidarDevice.Scan(context.Background(), lidar.ScanOptions{})
+		pc, err := cameraDevice.NextPointCloud(context.Background())
```
Doesn't seem to have a good reason; I assume it's a bug and will treat it as such.

kkufieta commented:
Work done:

- Fixed the mentioned issues
- Refactored the code a bit to make it more readable
- Added another sample file next to the client & server that can be used to run the rplidar and save files.

kkufieta commented:
PS: If you could pay close attention to my usage of `ctx` and help me make sure to pass it when it really is needed, that'd be great :) @Otterverse

jeremyhyde-viam reviewed (approved):
LGTM

Otterverse commented on cmd/savepcdfiles/main.go:
```diff
@@ -0,0 +1,106 @@
+
+		pc, err := rplidar.NextPointCloud(ctx)
+		if err != nil {
+			return multierr.Combine(ctx.Err(), myRobot.Close(ctx))
```
This isn't a context error, so combining that into the output is probably pointless (but also probably harmless.)

kkufieta replied:
ah, I think what this should probably be instead is this?

```go
return multierr.Combine(err, myRobot.Close(ctx))
```

Otterverse replied:
Good catch!

Otterverse commented on rplidar.go:
```diff
@@ -348,44 +365,76 @@ func (d *Device) scan(options lidar.ScanOptions) (lidar.Measurements, error) {
+}
+
+func (d *Device) savePCDFile(timeStamp time.Time, pc pointcloud.PointCloud) error {
+	f, err := os.Create("data/rplidar_data_" + timeStamp.Format("2006-01-02T15_04_05.0000") + ".pcd")
```
In the previous conversation, I mentioned just forcing UTC and then not using the zone in the filename, but I don't think my example actually showed setting it to UTC. If that matters (e.g. you might try to merge data from different timezones or daylight savings/standard) you might want to make this UTC first. Just add UTC() in there:

`timeStamp.UTC().Format(...)`

Up to you if that matters.

kkufieta replied:
It's a good detail to keep in mind. Thanks for clarifying it, and pointing it out! I'll add it.

Otterverse commented on rplidar.go:
```diff
@@ -348,44 +365,76 @@ func (d *Device) scan(options lidar.ScanOptions) (lidar.Measurements, error) {
+	if err = f.Close(); err != nil {
+		return err
+	}
+	return nil
```
[nit] You can slightly simplify these last few lines by directly returning f.Close() instead of checking if it's an error.

`return f.Close()`

Otterverse replied:
But otherwise, good catch to flush the buffer before closing the backing file too!

Otterverse reviewed (approved):
LGTM! A couple of minor things I noted, but I'd consider them optional. Nice work!
